rpc.go

```diff
@@ -60,65 +61,114 @@ func newRPCClient(ctx context.Context, ec entityConnector, opts ...rpcClientOpti
+		tab.For(ctx).Error(err)
+		return nil, err
+	}
 	return r, nil
 }
 
+// newClient will replace the existing client and restart auth auto-refresh.
+// any pre-existing client should be closed before calling this method.
+// NOTE: this does *not* take the write lock, callers must hold it as required!
+func (r *rpcClient) newClient(ctx context.Context) error {
+	var err error
+	r.client, err = r.ec.Namespace().newClient(ctx)
+	if err != nil {
+		return err
+	}
+	r.cancelAuthRefresh, err = r.ec.Namespace().negotiateClaim(ctx, r.client, r.ec.ManagementPath())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 // Recover will attempt to close the current session and link, then rebuild them
 func (r *rpcClient) Recover(ctx context.Context) error {
 	ctx, span := r.startSpanFromContext(ctx, "sb.rpcClient.Recover")
 	defer span.End()
-
-	_ = r.Close()
-	return r.ensureConn(ctx)
+	// atomically close and rebuild the client
+	r.clientMu.Lock()
+	defer r.clientMu.Unlock()
+	_ = r.close()
+	if err := r.newClient(ctx); err != nil {
+		tab.For(ctx).Error(err)
+		return err
+	}
+	return nil
 }
 
 // Close will close the AMQP connection
 func (r *rpcClient) Close() error {
 	r.clientMu.Lock()
 	defer r.clientMu.Unlock()
+	return r.close()
+}
+
+// closes the AMQP connection.  callers *must* hold the client write lock before calling!
+func (r *rpcClient) close() error {
 	if r.cancelAuthRefresh != nil {
 		<-r.cancelAuthRefresh()
 	}
-
 	return r.client.Close()
 }
 
-func (r *rpcClient) ensureConn(ctx context.Context) error {
-	ctx, span := r.startSpanFromContext(ctx, "sb.rpcClient.ensureConn")
-	defer span.End()
-
-	if r.client != nil {
-		return nil
+// creates a new link and sends the RPC request, recovering on failure
+func (r *rpcClient) doRPCWithRetry(ctx context.Context, address string, msg *amqp.Message, times int, delay time.Duration, opts ...rpc.LinkOption) (*rpc.Response, error) {
+	retries := 0
+	for {
+		r.clientMu.RLock()
+		client := r.client
+		r.clientMu.RUnlock()
+		var link *rpc.Link
+		var rsp *rpc.Response
+		var err error
+		link, err = rpc.NewLink(client, address, opts...)
+		if err == nil {
+			rsp, err = link.RetryableRPC(ctx, times, delay, msg)
+			if err == nil {
+				return rsp, err
+			}
+		}
+		if retries >= amqpRetryDefaultTimes || !isAMQPTransientError(err) {
+			return nil, err
+		}
+		// if we get here, something failed.  recover and try again
+		_ = r.Recover(ctx)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-time.After(amqpRetryDefaultDelay):
+			retries++
+		}
 	}
+}
 
-	r.clientMu.Lock()
-	defer r.clientMu.Unlock()
-
-	client, err := r.ec.Namespace().newClient(ctx)
-	r.cancelAuthRefresh, err = r.ec.Namespace().negotiateClaim(ctx, client, r.ec.ManagementPath())
-	if err != nil {
-		tab.For(ctx).Error(err)
-		_ = client.Close()
-		return err
+// returns true if the AMQP error is considered transient
+func isAMQPTransientError(err error) bool {
+	if errors.Is(err, amqp.ErrConnClosed) {
```

Comment:
I have mixed feelings on treating this error as transient.  Let's say the connection was intentionally closed (i.e. somebody called `Close()` on the queue).  This will cause us to effectively undo that, which would hide upstream bugs.

Reply:
This has been removed, and for the record, senders don't recover on a closed sender.

Reply:
Hmmm I do wonder about what happens when the connection was closed from the service side due to an error on their end? I believe that is why this was there.

Reply:
If it's closed due to a server error the `ErrConnClosed` won't be propagated to the caller so we'll attempt to recover the connection.

Reply:
To be more specific, in this case an `*amqp.Error` will be returned, and depending on the `Condition` we might or might not attempt to recover the connection.

Reply:
Ok I see what you're saying. Seems fine to remove it then.